lib/src/gainmapmath.cpp

```diff
@@ -739,4 +739,260 @@ uint64_t colorToRgbaF16(Color e_gamma) {
          (((uint64_t)floatToHalf(e_gamma.b)) << 32) | (((uint64_t)floatToHalf(1.0f)) << 48);
 }
 
+std::unique_ptr<uhdr_raw_image_ext_t> convert_raw_input_to_ycbcr(uhdr_raw_image_t* src) {
+  std::unique_ptr<uhdr_raw_image_ext_t> dst = nullptr;
+  float kYR, kYG, kYB, kCb, kCr;
+
+  if (src->fmt == UHDR_IMG_FMT_32bppRGBA1010102 || src->fmt == UHDR_IMG_FMT_32bppRGBA8888) {
+    if (src->cg == UHDR_CG_BT_709) {
+      kYR = kSrgbR;
+      kYG = kSrgbG;
+      kYB = kSrgbB;
+      kCb = kSrgbCb;
+      kCr = kSrgbCr;
+    } else if (src->cg == UHDR_CG_BT_2100) {
+      kYR = kBt2100R;
+      kYG = kBt2100G;
+      kYB = kBt2100B;
+      kCb = kBt2100Cb;
+      kCr = kBt2100Cr;
+    } else if (src->cg == UHDR_CG_DISPLAY_P3) {
+      kYR = kP3YR;
+      kYG = kP3YG;
+      kYB = kP3YB;
+      kCb = kP3Cb;
+      kCr = kP3Cr;
+    } else {
+      return dst;
+    }
+  }
+
+  if (src->fmt == UHDR_IMG_FMT_32bppRGBA1010102) {
+    dst = std::make_unique<uhdr_raw_image_ext_t>(UHDR_IMG_FMT_24bppYCbCrP010, src->cg, src->ct,
+                                                 UHDR_CR_LIMITED_RANGE, src->w, src->h, 64);
+
+    uint32_t* rgbData = static_cast<uint32_t*>(src->planes[UHDR_PLANE_PACKED]);
+    unsigned int srcStride = src->stride[UHDR_PLANE_PACKED];
+
+    uint16_t* yData = static_cast<uint16_t*>(dst->planes[UHDR_PLANE_Y]);
+    uint16_t* uData = static_cast<uint16_t*>(dst->planes[UHDR_PLANE_UV]);
+    uint16_t* vData = uData + 1;
+
+    for (size_t i = 0; i < dst->h; i += 2) {
+      for (size_t j = 0; j < dst->w; j += 2) {
+        float r0 = float(rgbData[srcStride * i + j] & 0x3ff);
+        float g0 = float((rgbData[srcStride * i + j] >> 10) & 0x3ff);
+        float b0 = float((rgbData[srcStride * i + j] >> 20) & 0x3ff);
+
+        float r1 = float(rgbData[srcStride * i + j + 1] & 0x3ff);
+        float g1 = float((rgbData[srcStride * i + j + 1] >> 10) & 0x3ff);
+        float b1 = float((rgbData[srcStride * i + j + 1] >> 20) & 0x3ff);
+
+        float r2 = float(rgbData[srcStride * (i + 1) + j] & 0x3ff);
+        float g2 = float((rgbData[srcStride * (i + 1) + j] >> 10) & 0x3ff);
+        float b2 = float((rgbData[srcStride * (i + 1) + j] >> 20) & 0x3ff);
+
+        float r3 = float(rgbData[srcStride * (i + 1) + j + 1] & 0x3ff);
+        float g3 = float((rgbData[srcStride * (i + 1) + j + 1] >> 10) & 0x3ff);
+        float b3 = float((rgbData[srcStride * (i + 1) + j + 1] >> 20) & 0x3ff);
+
+        r0 /= 1023.0f;
+        g0 /= 1023.0f;
+        b0 /= 1023.0f;
+
+        r1 /= 1023.0f;
+        g1 /= 1023.0f;
+        b1 /= 1023.0f;
+
+        r2 /= 1023.0f;
+        g2 /= 1023.0f;
+        b2 /= 1023.0f;
+
+        r3 /= 1023.0f;
+        g3 /= 1023.0f;
+        b3 /= 1023.0f;
+
+        float y0 = kYR * r0 + kYG * g0 + kYB * b0;
```

Comment:
Thank you for making this change. I think there is already color transfer methods in gainmapmath.cpp eg. `bt2100YuvToRgb` `bt2100RgbToYuv`, could you re-use them? Thanks!

Reply:
Done.